oxide-vpc/src/engine/icmpv6.rs

```diff
@@ -9,41 +9,169 @@
+// may be either any link-local address in guest (since we can't know how they
+// generate that) or its assigned VPC-private address. The destination address
+// must be the link-local address we derive for OPTE, from the EUI-64 transform
+// on its MAC address.
+//
+// - Respond to NDP Router Solicitations from the guest to the gateway.
+//
+// - Respond to NDP Neighbor Solicitations from the guest to the gateway, with a
+// Target Address of both the link-local and VPC-private IPv6 address.
+//
+// - Drop any other NDP traffic, inbound or outbound.
+//
+// - Pass through any other ICMPv6 traffic.
+fn add_icmpv6_rules(
+    pb: &mut PortBuilder,
+    cfg: &VpcCfg,
+    ip_cfg: &Ipv6Cfg,
+    ft_limit: NonZeroU32,
+) -> Result<(), OpteError> {
+    // We need to hairpin echo requests from either the VPC-private or
+    // link-local address of the guest, to OPTE's link-local.
+    let src_ips = [ip_cfg.private_ip, Ipv6Addr::from_eui64(&cfg.private_mac)];
+    let dst_ip = Ipv6Addr::from_eui64(&cfg.gateway_mac);
+    let n_pings = src_ips.len();
+    let mut actions = Vec::with_capacity(n_pings + 2);
+    for src_ip in src_ips.iter().copied() {
+        let echo = Action::Hairpin(Arc::new(Icmpv6EchoReply {
+            src_mac: cfg.private_mac,
+            src_ip,
+            dst_mac: cfg.gateway_mac,
+            dst_ip,
+        }));
+        actions.push(echo);
+    }
+
+    // Map an NDP Router Solicitation from the guest to a Router Advertisement
+    // from the OPTE virtual gateway's link-local IPv6 address.
+    let router_advert = Action::Hairpin(Arc::new(RouterAdvertisement::new(
+        // From the guest's private MAC.
+        cfg.private_mac,
+        // The MAC from which we respond, i.e., OPTE's MAC.
+        cfg.gateway_mac,
+        // "Managed Configuration", indicating the guest needs to use DHCPv6 to
+        // acquire an IPv6 address.
+        true,
+    )));
+    actions.push(router_advert);
+
+    // Map an NDP Neighbor Solicitation from the guest to a neighbor
+    // advertisement from the OPTE virtual gateway. Note that this is required
+    // perf RFC 4861 so that the guest does not mark the neighbor failed.
+    let neighbor_advert =
+        Action::Hairpin(Arc::new(NeighborAdvertisement::new(
+            // From the guest's private MAC.
+            cfg.private_mac,
+            // To OPTE's MAC.
+            cfg.gateway_mac,
+            // Set the ROUTER flag to true.
+            true,
+        )));
+    actions.push(neighbor_advert);
+
+    let n_actions = actions.len();
+    let mut icmp = Layer::new("icmpv6", pb.name(), actions, ft_limit);
+
+    // Add rules for the above actions.
+    for i in 0..n_actions {
+        let priority = u16::try_from(i + 1).unwrap();
+        let rule = Rule::new(priority, icmp.action(i).unwrap().clone());
+        icmp.add_rule(Direction::Out, rule.finalize());
+    }
+
+    // Drop any other NDP traffic in either direction, both in and out.
```

Comment:
So I think I'd approach this a different way.

First off, at the max priority, place a deny for **any** ICMPv6 traffic.

At `n_actions + 1` place an allow action (I think the static `Identity` action would do) for any ICMPv6 Echo/Reply. As we need other ICMPv6 traffic (like perhaps inbound DU) we can add it to the list of allowed traffic.

Reply:
Yeah, that's nice. I'll update it now, thanks.

Reply:
I'm not sure I follow the Echo Request / Reply bit, actually. I think you're recommending just allowing _all_ ICMPv6 traffic, regardless of destination. I assume that means you're expecting the router layer to filter those on the basis of the source / destination IP addresses, is that right?

Reply:
No, basically I'm saying to create an allowed list of ICMP types as opposed to your deny list. I think it would look something like the following (note I didn't actually verify this compiles, might not be 100% right).


```Rust
let mut echo = Rule::new(nactions + 1, Action::Static(Identity::new());
echo.add_predicate(Predicate::InnerEtherType(vec![EtherTypeMatch::Exact(
     ETHER_TYPE_IPV6,
 )]));
 echo.add_predicate(Predicate::InnerIpProto(vec![IpProtoMatch::Exact(
     Protocol::ICMPv6,
 )]));
// Actually it would be nice to make this a Vec of message types so you could just apply this rule to both echo and reply.
echo.add_data_predicate(DataPredicate::Icmpv6MsgType(MessageType::from(
     Icmpv6Message::Echo,
 )));
icmp.add_rule(Direction::Out, echo.clone().finalize());
icmp.add_rule(Direction::In, echo.finalize());

let deny = Rule::new(u16::MAX - 1, Action::Deny);
deny.add_predicate(Predicate::InnerEtherType(vec![EtherTypeMatch::Exact(
     ETHER_TYPE_IPV6,
 )]));
 deny.add_predicate(Predicate::InnerIpProto(vec![IpProtoMatch::Exact(
     Protocol::ICMPv6,
 )]));
icmp.add_rule(Direction::Out, deny.clone().finalize());
imcp.add_rule(Direction::Out, deny.finalize());
```

Reply:
So that all makes sense, and that's what I attempted to do in the follow-up commit. (Might have mucked that up, though.) I think I was trying to clarify whether there are predicates here on the destination or source IP addresses. For example, should we deny inbound ICMPv6 that's not destined for the client IP addresses? Or is the expectation that that denial would happen at the router or firewall?

Does that make sense?

Reply:
BTW, I see the confusion. I wrote "allowing _all_ ICMPv6 traffic" in my reply. I meant "denying _all_ ICMPv6 traffic". So...yeah, that would be confusing :)

Reply:
Okay, yea that typo had me very confused. And yes, I'm saying **deny all** ICMPv6 traffic that doesn't match the allow list.

The router doesn't even do any checks on inbound (maybe something to change in the future but I don't want to rat hole on that right now).

The firewall is user controlled and is separate from OPTE's own enforcement.

In this case the ICMP layer should make sure only the explicit ICMP traffic we allow is allowed to pass. For now let's just keep that to echo/reply (but we will probably need inbound DU support soonish).

There is still the larger issue of #42, where we make sure early that the guest doesn't have some malicious program using a raw socket/ip/L2 API to send out bogus traffic. That should mostly just be an outbound concern I think, but also separate from this PR as it applies to all traffic, not just ICMP. Though there is no harm tacking on the extra predicates to the rules in the this layer as well to make sure things are coming from the expected MAC address and IP.